library/Booster/Pattern/Base.hs

```diff
@@ -252,21 +291,112 @@ internaliseKmap def@KMapDefinition{symbolNames = KMapAttributes{elementSymbolNam
                             , hash =
                                 Hashable.hash
                                     ( "SymbolApplication" :: ByteString
-                                    , kmapConcatSymbol def
+                                    , concatSym
                                     , map hash [aAttribs, bAttribs]
                                     )
                             , isConstructorLike = False
                             }
-                 in Just $ Term attribs $ SymbolApplicationF (kmapConcatSymbol def) [] [a, b]
+                 in Just $ Term attribs $ SymbolApplicationF concatSym [] [a, b]
         )
 
+{- | try to represent a list-returning symbol application as an
+  internal list with head, optional middle and optional tail (empty
+  list otherwise).
+
+  In case of a concatenation, we might not succeed in representing the
+  result as the internal list type, and will construct the result
+  using list concatenation instead.
+-}
+internaliseKList :: KListDefinition -> Term -> Term
+internaliseKList def = \case
+    SymbolApplication s _ []
+        | s.name == def.symbolNames.unitSymbolName -> KList def [] Nothing
+    SymbolApplication s _ [x]
+        | s.name == def.symbolNames.elementSymbolName -> KList def [x] Nothing
+    SymbolApplication concatSym _ [x, y]
+        | concatSym.name == def.symbolNames.concatSymbolName ->
+            case (internaliseKList def x, internaliseKList def y) of
+                -- try to combine cases that can be represented as `heads mid tails`
+                (KList def1 hds1 rst1, KList def2 hds2 rst2)
+                    | def1 /= def2 -> inconsistent def1 def2
+                    | def /= def1 -> inconsistent def def1
+                    | Nothing <- rst1
+                    , Nothing <- rst2 ->
+                        KList def1 (hds1 <> hds2) Nothing
+                    | Nothing <- rst1 ->
+                        KList def1 (hds1 <> hds2) rst2
+                    | Nothing <- rst2 ->
+                        KList def1 hds1 $ fmap (second (<> hds2)) rst1
+                -- otherwise neither mid1 nor mid2 are trivial, we
+                -- reconstruct concat expression.
+                (a@KList{}, b@KList{}) ->
+                    let attribs = concatAttribs (getAttributes a) (getAttributes b)
+                     in Term attribs $ SymbolApplicationF concatSym [] [a, b]
```

Comment:
Perhaps we should have

```haskell
data TermF t
    = ...
    | -- | internal List
      KListF
        KListDefinition -- metadata
        [t] -- head elements
        [(t, [t])]
```

And merging
`KList e_1:...e_n [...,(MID, e_l..:e_k)]` with `KList e'_1:...e'_n rs` becomes

`KList e_1:...e_n ([...,(MID, e_l..:e_k ++ e'_1:...e'_n)] ++ rs)`

Reply:
perhaps best left for next iteration if/when we want to make the unification more powerful

Reply:
yes, this representation would allow for removing _every_ `_List_` (concat) application . But matching / unification will probably be more tricky.